### ISSUE TITLE
Memory leak when using generators

### ISSUE THREAD

**Issue by prapin** [philsquared/Catch#440]
CATCH has a memory leak when using generators (`GENERATE` macro).
It can be easily seen using Valgrind or MSVC (refer issue #439) with a single unit test calling `GENERATE`.
My proposed fix is to maintain a vector of generators in `Context` to delete them.

See my commit for reference:
https://github.com/prapin/Catch/commit/46ba712e2c3aa6c8c055709d6001e04577cafdf0

**Comment by DavidEGrayson** [comment]
I too think there might be a memory leak in Catch somewhere.  When I run my tests with [Dr. Memory](http://www.drmemory.org/), the output of Dr. Memory says:

Error #1: POSSIBLE LEAK 12 direct bytes 0x02fabb80-0x02fabb8c + 0 indirect bytes
# 0 replace_malloc                                        [d:\drmemory_package\common\alloc_replace.c:2377]
# 1 libgcc_s_dw2-1.dll!?                                 +0x0      (0x6eb550be <libgcc_s_dw2-1.dll+0x150be>)
# 2 libstdc++-6.dll!?                                    +0x0      (0x6feb827f <libstdc++-6.dll+0x7827f>)
# 3 Catch::Section::~Section                              [C:/msys64/mingw32/include/catch.hpp:7085]
# 4 ____C_A_T_C_H____T_E_S_T____5                         [C:/.......cpp:15]
# 5 Catch::FreeFunctionTestCase::invoke                   [C:/msys64/mingw32/include/catch.hpp:5818]
# 6 Catch::TestCase::invoke                               [C:/msys64/mingw32/include/catch.hpp:6723]
# 7 Catch::RunContext::invokeActiveTestCase               [C:/msys64/mingw32/include/catch.hpp:5417]
# 8 Catch::RunContext::runCurrentTest                     [C:/msys64/mingw32/include/catch.hpp:5389]
# 9 Catch::RunContext::runTest                            [C:/msys64/mingw32/include/catch.hpp:5228]
#10 Catch::Runner::runTests                               [C:/msys64/mingw32/include/catch.hpp:5547]
#11 Catch::Session::run                                   [C:/msys64/mingw32/include/catch.hpp:5678]

Sorry that this isn't a complete bug report with a minimal test case.  Just thought I would mention it in case it is useful.

**Comment by prapin** [comment]
Sorry if I wasn't clear.
The minimum test case is trivial:

```C++
TEST_CASE("generator", "generator")
{
	using namespace Catch::Generators;
	int a = GENERATE(between(0, 1));
}
```

And the memory leak is clearly identified. The generator object created by `createGeneratorsForTest` is never deleted. The commit I put in reference fixes the leak.

**Comment by philsquared** [comment]
The current implementation of generators is not complete - and should be considered deprecated (I've had to write a whole new infrastructure for it to work correctly with sections - will have something to push some time in the future).

**Comment by horenmar** [comment]
This was fixed in c65aeaf25f63de7aa95dffef7fe8ddb7434cc5a9
